src/bots/userPnlSettler.ts

```diff
@@ -506,254 +439,459 @@ export class UserPnlSettlerBot implements Bot {
-					const usersChunk = sortedParams.slice(i, i + SETTLE_USER_CHUNKS);
-					allTxPromises.push(this.trySendTxForChunk(marketIndex, usersChunk));
-				}
+		let settleePositionWithLp = settleePosition;
+		if (!settleePosition.lpShares.eq(ZERO)) {
+			settleePositionWithLp = user.getPerpPositionWithLPSettle(
+				perpMarketIdx,
+				settleePosition
+			)[0];
+		}
 
-				logger.info(`Waiting for ${allTxPromises.length} txs to settle...`);
-				const settleStart = Date.now();
-				await Promise.all(allTxPromises);
-				logger.info(
-					`Settled ${sortedParams.length} users in market ${marketIndex} in ${
-						Date.now() - settleStart
-					}ms`
-				);
+		if (
+			settlePnlWithPositionPaused &&
+			!settleePositionWithLp.baseAssetAmount.eq(ZERO)
+		) {
+			logger.debug(
+				`Skipping user ${userAccKeyStr}-${settleePosition.marketIndex} because settling pnl with position blocked`
+			);
+			return { shouldSettle: false };
+		}
+
+		// Check market data availability
+		if (
+			!marketDataCache.perpMarketAndOracleData[perpMarketIdx] ||
+			!marketDataCache.spotMarketAndOracleData[spotMarketIdx]
+		) {
+			logger.error(
+				`Skipping user ${userAccKeyStr}-${settleePosition.marketIndex} because no spot market or oracle data`
+			);
+			return { shouldSettle: false };
+		}
+
+		const userUnsettledPnl = calculateClaimablePnl(
+			marketDataCache.perpMarketAndOracleData[perpMarketIdx].marketAccount,
+			marketDataCache.spotMarketAndOracleData[spotMarketIdx].marketAccount,
+			settleePositionWithLp,
+			marketDataCache.perpMarketAndOracleData[perpMarketIdx].oraclePriceData
+		);
+
+		// Check LP settlement conditions
+		const shouldSettleLp = await this.shouldSettleLpPosition(
+			settleePosition,
+			perpMarket,
+			nowTs,
+			marketDataCache.perpMarketAndOracleData[perpMarketIdx]
+		);
+
+		if (settleePositionWithLp.lpShares.gt(ZERO) && !shouldSettleLp) {
+			logger.debug(
+				`Skipping user ${userAccKeyStr}-${settleePosition.marketIndex} with lpShares and shouldSettleLp === false`
+			);
+			return { shouldSettle: false };
+		}
+
+		// Apply PnL filtering based on options
+		if (options.positiveOnly) {
+			if (userUnsettledPnl.lte(ZERO)) {
+				return { shouldSettle: false };
 			}
-		} catch (err) {
-			console.error('Error in trySettlePnl', err);
-			if (!(err instanceof Error)) {
-				return;
+			// For positive PnL, check if we can settle it
+			const canSettlePositivePnl = await this.canSettlePositivePnl(
+				user,
+				userUnsettledPnl,
+				perpMarketIdx,
+				spotMarketIdx,
+				marketDataCache
+			);
+			if (!canSettlePositivePnl) {
+				return { shouldSettle: false };
 			}
+		} else {
+			// For negative PnL settlement, apply existing logic
 			if (
-				!err.message.includes('Transaction was not confirmed') &&
-				!err.message.includes('Blockhash not found')
+				(userUnsettledPnl.eq(ZERO) &&
+					settleePositionWithLp.lpShares.eq(ZERO)) ||
+				(userUnsettledPnl.gt(this.minPnlToSettle) &&
```

Comment:
hmm....i think this needed to be abs

```suggestion
				(userUnsettledPnl.abs().gt(this.minPnlToSettle) &&
```